nixos.nix

```diff
@@ -88,15 +88,66 @@ in
     system.activationScripts =
       let
         # Create a directory in persistent storage, so we can bind
-        # mount it.
-        mkDirCreationSnippet = persistentStoragePath: dir:
-          let
-            targetDir = concatPaths [ persistentStoragePath dir ];
-          in
+        # mount it. The directory structure's mode and ownership mirror those of
+        # persistentStoragePath/dir;
+        # TODO: Move this script to it's own file, add CI with shfmt/shellcheck.
+        mkDirWithModeAndPerms = persistentStoragePath: dir:
           ''
-            if [[ ! -e "${targetDir}" ]]; then
-                mkdir -p "${targetDir}"
+            # Given a source directory, /source, and a target directory,
+            # /target/foo/bar/bazz, we want to "clone" the target structure
+            # from source into the target. Essentially, we want both
+            # /source/target/foo/bar/bazz and /target/foo/bar/bazz to exist
+            # on the filesystem. More concretely, we'd like to map
+            # /state/etc/ssh/example.key to /etc/ssh/example.key
+            #
+            # To achieve this, we split the target's path into parts -- target, foo,
+            # bar, bazz -- and iterate over them while accumulating the path
+            # (/target/, /target/foo/, /target/foo/bar, and so on); then, for each of
+            # these increasingly qualified paths we:
+            #
+            #   1. Ensure both /source/qualifiedPath and qualifiedPath exist
+            #   2. Copy the ownership of the source path into the target path
+            #   3. Copy the mode of the source path into the target path
+            (
+            # capture the nix vars into bash to avoid escape hell
+            sourceBase="${persistentStoragePath}"
+            target="${dir}"
+
+            # trim trailing slashes the root of all evil
+            sourceBase="''${sourceBase%/}"
+            target="''${target%/}"
+
+            # check that the source exists, if it doesn't exit the underlying
+            # scope (the activation script will continue)
+            realSource="$(realpath "$sourceBase$target")"
+            if [ ! -d "$realSource" ]; then
+                printf "\e[1;31mBind source '%s' does not exist!\e[0m\n" "$realSource"
+                exit 1
```

Comment:
Wait, why fail instead of just creating the directory? :confused:

Reply:
True, I've reverted this. It now warns if the source doesn't exist, but continues to create it.